### DISCUSSION TITLE
Noisy results with resampling (aliasing?)

### DISCUSSION THREAD

**Discussion by KontinuumLab** [pschatzmann/arduino-audio-tools#1330] [Q&A] [answered]
I'm working on an instrument which plays a sample of a Hang instrument on an ESP32, out through a PCM5102 DAC.
in the test sketch pasted below I am resampling the original sample, to get the whole series of notes I need, using a series of multiplication factors meant to reproduce semitones, from one octave lower and to one octave higher, which works quite well. The multiplication factors can be found in the "freqMult" array. 
The problem is that most of the resampled notes are slightly distorted, producing a variety of buzzing noises over the actual sample, which is what I think is commonly called "aliasing". The only frequencies that sound completely clean out of my whole array are those produced by factors of 0.5, 1.0, 1.5 and 2.0.
I haven't included the actual sample, as it's quite a large array. I'm happy to share it though, if someone wants to test this.
Is there another way of doing this, which avoids the aliasing?
Thanks!

```
#include "AudioTools.h"
#include "hang_F.h"

float freqMult[25] = {0.50000, 0.52973, 0.56123, 0.59460, 0.62995, 0.66741, 0.70709, 0.74914, 0.79368, 0.84087, 0.89087, 0.94384, 
                      1.0000, 1.05946, 1.12245, 1.18919, 1.25990, 1.33482, 1.41418, 1.5000, 1.58919, 1.68368, 1.78380, 1.88985, 2.0000};
AudioInfo info(22050, 1, 16);
I2SStream i2s;  // Output to I2S
MemoryStream hang_F(hang_F_raw, hang_F_raw_len);
ResampleStream resample(i2s);
StreamCopy copier(resample, hang_F); // copies sound into i2s

unsigned long timer = 0;
int justStarted = 1;
float stepSize = 0.5;
int pos = 0;

void setup(){
    Serial.begin(115200);
    // AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // define resampling info
  auto rcfg = resample.defaultConfig();
  rcfg.copyFrom(info);
  rcfg.step_size = stepSize;
  resample.begin(rcfg);

  auto ocfg = i2s.defaultConfig(TX_MODE);
//  config.copyFrom(info);
  ocfg.bits_per_sample = 16;
  ocfg.sample_rate = 22050;
  ocfg.channels = 1;
  ocfg.pin_bck = 26;
  ocfg.pin_ws = 25;
  ocfg.pin_data = 22;

    i2s.begin(ocfg);

}

void loop(){
    // if (!copier.copy()){
    //   i2s.end();
    //   stop();
    // }
  copier.copy();

  if(millis() >= timer + 1000){
    Serial.println(freqMult[pos], 5);
    timer = millis();
    auto rcfg = resample.defaultConfig();
    stepSize = freqMult[pos];
    rcfg.step_size = stepSize;
    resample.begin(rcfg);
    hang_F.begin();
    pos = pos += 1;
    if(pos > 24){
      pos = 0;
    }
  }
}
```

**Comment by pschatzmann** [comment]
I am not sure, can you provide me with some recording of the sound ?
Ps to calculate the frequencies you could also rely on [this api](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_musical_notes.html)

**Comment by KontinuumLab** [comment]
Here's a recording of the full sequence. (zipped, because github won't accept the mp3)
As you can hear, notes 1, 13, 20 and 25 are clean. All the rest have some aliasing.

[Hang_AudioTest_Aliasing_1.zip](https://github.com/pschatzmann/arduino-audio-tools/files/14270475/Hang_AudioTest_Aliasing_1.zip)

**Comment by pschatzmann** [comment, marked as answer]
Not sure what the easiest approach would be. 
It might help if you run the output thru a Median filter. 

I found the following with the help of Google. The recommended way of converting an oversampled bandlimited signal into its critical sampling rate (or somehow above that) is to use a time-domain Low Pass filter and decimate approach.

**Comment by KontinuumLab** [comment]
Thanks Phil. That sounds promising.
I did some research as well and found that to avoid aliasing  _"the sampling frequency must be at least twice that of the highest frequency which is to be represented"._ So I decided to up the sampling frequency to 44100, which has helped quite a lot. It's still not great, but better than before. (New audio sample attached)
[Hang_AudioTest_Aliasing_2.zip](https://github.com/pschatzmann/arduino-audio-tools/files/14271663/Hang_AudioTest_Aliasing_2.zip)
Now, as for the filtering you suggest, I'm familiar with median filtering for sensor input, but not sure what it would look like in an audio context. 
Is there a suitable filter in the audio-tools?

**Comment by pschatzmann** [comment]
[MedianFilter](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_median_filter.html)<int16_t> medianFilter;

**Comment by KontinuumLab** [comment]
Looking at the link you posted, I find that the MedianFilter is supposed to be included in the library, at: "src/AudioFilter/MedianFilter.h"
However, in my audio-tools, the filter folder looks like this: 
![audioFilter_Folder](https://github.com/pschatzmann/arduino-audio-tools/assets/10105083/6bef14de-5959-47d6-abd6-826597dd121b)
So, no MedianFilter... So then I tried to simply make a new .h file with the filter function and include it with my sketch and use the filter from there, but I'm afraid I have no idea how to implement it, and so far I just get errors.
Sorry if I'm being dense here, but could you give me some pointers as to how to use it?

**Comment by pschatzmann** [comment]
Maybe you have an old version of the library. It is definitely [there](https://github.com/pschatzmann/arduino-audio-tools/tree/main/src/AudioFilter)

**Comment by KontinuumLab** [comment]
Ok, was busy for a couple of days, but now I'm back at it.
I downloaded the new library, including the MedianFilter, but I can't figure out how to implement it in my sketch (still the test sketch posted in the original question). 
Sorry for being a bit dense here, but could you show me where to insert it in the chain, and how?
Thanks!

**Comment by pschatzmann** [comment]
There is a [Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/Filters) that explains how to use filters and there is even a [test sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/tests/filters/test-median-filter/test-median-filter.ino) using mono audio

**Reply by KontinuumLab** [reply]
Ok thanks!
